src/apps/js_generic/js_generic.cpp

```diff
@@ -436,6 +436,57 @@ namespace ccfapp
     return JS_ThrowTypeError(ctx, "Cannot call set on read-only map");
   }
 
+  static JSValue js_kv_map_foreach(
+    JSContext* ctx, JSValueConst this_val, int argc, JSValueConst* argv)
+  {
+    auto map_view =
+      static_cast<KVMap::TxView*>(JS_GetOpaque(this_val, kv_map_view_class_id));
+
+    if (argc != 1)
+      return JS_ThrowTypeError(
+        ctx, "Passed %d arguments, but expected 1", argc);
+
+    JSValue func = argv[0];
+
+    if (!JS_IsFunction(ctx, func))
+      return JS_ThrowTypeError(ctx, "Argument must be a function");
+
+    bool failed = false;
+    map_view->foreach(
+      [ctx, this_val, func, &failed](const auto& k, const auto& v) {
+        JSValue args[3];
+
+        // JS forEach expects (v, k, map) rather than (k, v)
+        args[0] = JS_NewArrayBufferCopy(ctx, v.data(), v.size());
```

Comment:
I don't think we need copies here, shouldn't `JS_NewArrayBuffer` work as well?

Reply:
The problem is that the JS code might make copies of the ArrayBuffer (used entirely outside of the `forEach`), so to support that we need to extend the lifetime of the data and provide a function to eventually free it. The data in `k` and `v` is only safe for the duration of this C++ callback.

Reply:
Got it, ok, I don't see a way around a copy then.